x/ibc/02-client/client/rest/query.go

```diff
@@ -0,0 +1,185 @@
+		if !ok {
+			return
+		}
+
+		req := abci.RequestQuery{
+			Path:  "store/ibc/key",
+			Data:  types.KeyClientState(clientID),
+			Prove: rest.ParseQueryProve(r),
+		}
+
+		res, err := cliCtx.QueryABCI(req)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		var state types.State
+		if err := cliCtx.Codec.UnmarshalBinaryLengthPrefixed(res.Value, &state); err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+		}
+
+		cliCtx = cliCtx.WithHeight(res.Height)
+		rest.PostProcessResponse(w, cliCtx, types.NewClientStateResponse(clientID, state, res.Proof, res.Height))
+	}
+}
+
+func queryRootHandlerFn(cliCtx context.CLIContext, queryRoute string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		clientID := vars[RestClientID]
+		height, err := strconv.ParseUint(vars[RestRootHeight], 10, 64)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		cliCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, cliCtx, r)
+		if !ok {
+			return
+		}
+
+		bz, err := cliCtx.Codec.MarshalJSON(types.NewQueryCommitmentRootParams(clientID, height))
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		res, resHeight, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s", queryRoute, types.QueryVerifiedRoot), bz)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		cliCtx = cliCtx.WithHeight(resHeight)
+		rest.PostProcessResponse(w, cliCtx, res)
+	}
+}
+
+func queryNodeConsensusStateHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		node, err := cliCtx.GetNode()
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		info, err := node.ABCIInfo()
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		height := info.Response.LastBlockHeight
+		prevHeight := height - 1
+
+		commit, err := node.Commit(&height)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		validators, err := node.Validators(&prevHeight)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		state := tendermint.ConsensusState{
+			ChainID:          commit.ChainID,
+			Height:           uint64(commit.Height),
+			Root:             commitment.NewRoot(commit.AppHash),
+			NextValidatorSet: tmtypes.NewValidatorSet(validators.Validators),
+		}
+
+		res := cliCtx.Codec.MustMarshalJSON(state)
+		cliCtx = cliCtx.WithHeight(height)
+		rest.PostProcessResponse(w, cliCtx, res)
+	}
+}
+
+func queryPathHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
```

Comment:
I see that this is hardcoded here. We can remove it and move it inside the code where it's needed for a better UX. Same applies to the CLI

Reply:
Agree, how about moving to `keys.go` and providing a Querier method？